commit 477f0425602d4613fdf8370ab2be4716b7cce5a5
Author: Kouhei Sutou <kou@clear-code.com>
Date:   Wed Jun 12 17:43:40 2013 +0900

    Add Operator::AND_NOT and deprecate Operator::NOT
    
    Because NOT is deprecated in groonga.

diff --git a/ext/groonga/rb-grn-column.c b/ext/groonga/rb-grn-column.c
index 37ab76c..6c669f6 100644
--- a/ext/groonga/rb-grn-column.c
+++ b/ext/groonga/rb-grn-column.c
@@ -228,7 +228,7 @@ rb_grn_column_get_local_name (VALUE self)
  *       - Groonga::Operator::AND :=
  *         マッチしたレコードのスコアを増加。マッチしなかった
  *         レコードを削除。 =:
- *       - Groonga::Operator::BUT :=
+ *       - Groonga::Operator::AND_NOT :=
  *         マッチしたレコードを削除。 =:
  *       - Groonga::Operator::ADJUST :=
  *         マッチしたレコードのスコアを増加。 =:
diff --git a/ext/groonga/rb-grn-double-array-trie.c b/ext/groonga/rb-grn-double-array-trie.c
index 0cfcb2c..fe90a61 100644
--- a/ext/groonga/rb-grn-double-array-trie.c
+++ b/ext/groonga/rb-grn-double-array-trie.c
@@ -286,7 +286,7 @@ rb_grn_double_array_trie_s_create (int argc, VALUE *argv, VALUE klass)
  *     - Groonga::Operator::AND :=
  *       マッチしたレコードのスコアを増加。マッチしなかった
  *       レコードを削除。 =:
- *     - Groonga::Operator::BUT :=
+ *     - Groonga::Operator::AND_NOT :=
  *       マッチしたレコードを削除。 =:
  *     - Groonga::Operator::ADJUST :=
  *       マッチしたレコードのスコアを増加。 =:
diff --git a/ext/groonga/rb-grn-expression.c b/ext/groonga/rb-grn-expression.c
index 706f8dd..dd74c29 100644
--- a/ext/groonga/rb-grn-expression.c
+++ b/ext/groonga/rb-grn-expression.c
@@ -277,7 +277,7 @@ rb_grn_expression_append_operation (VALUE self, VALUE rb_operation,
  *       レコードはいずれかの条件にマッチすればいい。 =:
  *     - Groonga::Operator::AND :=
  *       レコードは全ての条件にマッチしなければならない。 =:
- *     - Groonga::Operator::BUT :=
+ *     - Groonga::Operator::AND_NOT :=
  *       最初の条件にレコードはマッチし、残りの条件にレコードは
  *       マッチしてはならない。 =:
  *
diff --git a/ext/groonga/rb-grn-operator.c b/ext/groonga/rb-grn-operator.c
index 011b833..2e534ff 100644
--- a/ext/groonga/rb-grn-operator.c
+++ b/ext/groonga/rb-grn-operator.c
@@ -35,15 +35,19 @@ rb_grn_operator_from_ruby_object (VALUE rb_operator)
         operator = GRN_OP_AND;
     } else if (rb_grn_equal_option(rb_operator, "but") ||
                rb_grn_equal_option(rb_operator, "not") ||
-               rb_grn_equal_option(rb_operator, "-")) {
-        operator = GRN_OP_BUT;
+               rb_grn_equal_option(rb_operator, "and-not") ||
+               rb_grn_equal_option(rb_operator, "and_not") ||
+               rb_grn_equal_option(rb_operator, "-") ||
+               rb_grn_equal_option(rb_operator, "&!")) {
+        operator = GRN_OP_AND_NOT;
     } else if (rb_grn_equal_option(rb_operator, "adjust") ||
                rb_grn_equal_option(rb_operator, ">")) {
         operator = GRN_OP_ADJUST;
     } else {
         rb_raise(rb_eArgError,
                  "operator should be one of "
-                 "[:or, :||, :and, :+, :&&, :but, :not, :-, :adjust, :>]: <%s>",
+                 "[:or, :||, :and, :+, :&&, :but, :not, :and_not, :-, :&!, "
+                 ":adjust, :>]: <%s>",
                  rb_grn_inspect(rb_operator));
     }
 
@@ -73,6 +77,9 @@ rb_grn_init_operator (VALUE mGrn)
                     UINT2NUM(GRN_OP_GET_VALUE));
     rb_define_const(rb_mGrnOperator, "AND",
                     UINT2NUM(GRN_OP_AND));
+    rb_define_const(rb_mGrnOperator, "AND_NOT",
+                    UINT2NUM(GRN_OP_AND_NOT));
+    /* Just for backward compatiblity. TODO: REMOVE ME! */
     rb_define_const(rb_mGrnOperator, "BUT",
                     UINT2NUM(GRN_OP_BUT));
     rb_define_const(rb_mGrnOperator, "OR",
diff --git a/ext/groonga/rb-grn-patricia-trie.c b/ext/groonga/rb-grn-patricia-trie.c
index 4437a2e..4bcb036 100644
--- a/ext/groonga/rb-grn-patricia-trie.c
+++ b/ext/groonga/rb-grn-patricia-trie.c
@@ -274,7 +274,7 @@ rb_grn_patricia_trie_s_create (int argc, VALUE *argv, VALUE klass)
  *     - Groonga::Operator::AND :=
  *       マッチしたレコードのスコアを増加。マッチしなかった
  *       レコードを削除。 =:
- *     - Groonga::Operator::BUT :=
+ *     - Groonga::Operator::AND_NOT :=
  *       マッチしたレコードを削除。 =:
  *     - Groonga::Operator::ADJUST :=
  *       マッチしたレコードのスコアを増加。 =:
diff --git a/ext/groonga/rb-grn-table.c b/ext/groonga/rb-grn-table.c
index cb82430..2959e0d 100644
--- a/ext/groonga/rb-grn-table.c
+++ b/ext/groonga/rb-grn-table.c
@@ -2005,7 +2005,7 @@ rb_grn_table_is_locked (int argc, VALUE *argv, VALUE self)
  *         は何もしない。 =:
  *       - Groonga::Operator::AND :=
  *         マッチしたレコードのスコアを増加。マッチしなかったレコードを削除。 =:
- *       - Groonga::Operator::BUT :=
+ *       - Groonga::Operator::AND_NOT :=
  *         マッチしたレコードを削除。 =:
  *       - Groonga::Operator::ADJUST :=
  *         マッチしたレコードのスコアを増加。 =:
@@ -2218,7 +2218,7 @@ rb_grn_table_intersection_bang (VALUE self, VALUE rb_other)
 static VALUE
 rb_grn_table_difference_bang (VALUE self, VALUE rb_other)
 {
-    return rb_grn_table_set_operation_bang(self, rb_other, GRN_OP_BUT);
+    return rb_grn_table_set_operation_bang(self, rb_other, GRN_OP_AND_NOT);
 }
 
 /*
